### PR TITLE
Feature/faith/campaign api and location multiple image upload

### PR DESCRIPTION
**Pull request by codemutatio**
Campaign API and uploading multiple images to the location

### CODE REVIEW

File: functions/src/controllers/campaign.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            const { valid, errors } = await validateCampaignInput(req.body);
+            if (!valid) validationError(res, errors);
+            req.body.createdAt = new Date().toISOString();
+            req.body.createdBy = req.user.uid;
+            let amount = await getLocationsAmount(req.body.locationsSelected);
+            amount = amount.reduce((a, b) => a + b, 0);
+            req.body.amount = amount * req.body.duration;
+            if (valid) {
+                await db.collection('campaigns').doc().create(req.body)
+                    .then(
```
Comment (enyeInc) [inline functions/src/controllers/campaign.js]: what is the purpose of this `.then`

File: functions/src/controllers/campaign.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        } catch (error) {
+            tryCatchError(res, error);
+        }
+    },
+
+    async getAll(req, res) {
+        try {
+            const campaigns = [];
+            const { uid } = req.user;
+            await db.collection('campaigns').where('createdBy', '==', uid)
```
Comment (enyeInc) [inline functions/src/controllers/campaign.js]: either use `await` or `.then` not both

File: functions/src/utils/functions.js [hunk 3]
```diff
@@ -51,6 +51,35 @@ const updateVideo = async (videoId, filePath, { description, title }) => {
...
+    const promises = [];
+    images.forEach(image => {
+        promises.push(getFirebaseLink(image.originalname, token));
+    });
+    return await Promise.all(promises);
+}; 
 module.exports = {
-    createUserData, getFirebaseLink, superAdmin, updateVideo, uploadRequest,
+    createUserData, getFirebaseLink, getLocationsAmount, getMultipleFirebaseLink, superAdmin, 
+    updateVideo, uploadMultipleImages, uploadRequest,
```
Comment (enyeInc) [inline functions/src/utils/functions.js:77]: each on their own line

File: functions/src/validations/campaignInput.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * An helper function to validate campaign input
+ * @function
+ * @param [array] datas - data to validata
+ * @return  {object} result
+ */
+const validateCampaignInput = datas => {
+    const errors = {};
+    const expectedData = [ commercialId, duration, locationsSelected, status, title, validity ];
+    const key = arrayCompare(expectedData, Object.keys(datas));
```
Comment (enyeInc) [inline functions/src/validations/campaignInput.js:25]: can you explain what is happening from line 24 - 25?

File: functions/src/utils/functions.js [hunk 5]
```diff
@@ -51,6 +51,35 @@ const updateVideo = async (videoId, filePath, { description, title }) => {
...
+        promises.push(uploadRequest(image, token));
+
+    });
+    return await Promise.all(promises);
+};
+const getMultipleFirebaseLink = async (images, token) => {
+    const promises = [];
+    images.forEach(image => {
+        promises.push(getFirebaseLink(image.originalname, token));
+    });
```
Comment (enyeInc) [inline functions/src/utils/functions.js]: Use this pattern for the above util functions as well
```
const promises = images.map(image => getFirebaseLink(image.originalname, token));
```

File: functions/src/controllers/savedLocation.js [hunk 6]
```diff
@@ -50,16 +49,18 @@ const SavedLocations = {
...
-                for (const doc of docs) {
-                    const selectedItem = {
-                        id: doc.id,
-                        savedLocation: doc.data(),
-                    };
-                    savedLocations.push(selectedItem);
-                } return savedLocations;
-            });
+            const { uid } = req.user;
+            await db.collection('savedLocations').where('createdBy', '==', uid )
```
Comment (enyeInc) [inline functions/src/controllers/savedLocation.js]: use await or then, not both

File: functions/src/controllers/campaign.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    async updateCampaign(req, res) {
+        try {
+            const document = db.collection('campaigns').doc(req.params.id);
+            if (!document) validationError(res, errors);
+            const { valid, errors } = await validateCampaignInput(req.body);
+            if (!valid) validationError(res, errors);
+            req.body.updatedAt = new Date().toISOString();
+            req.body.updatedBy = req.user.uid;
+            let amount = await getLocationsAmount(req.body.locationsSelected);
+            amount = amount.reduce((a, b) => a + b, 0);
```
Comment (enyeInc) [inline functions/src/controllers/campaign.js:96]: getLocationsAmount should probably return the reduced array

File: functions/src/validations/campaignInput.js (hunk 4 above)
Comment (codemutatio) [reply functions/src/validations/campaignInput.js:25]: This just taking an array of expected keys for the campaign input and check if what is inputted campaign data is as expected

File: functions/src/utils/functions.js (hunk 5 above)
Comment (codemutatio) [reply functions/src/utils/functions.js]: ok i will now

File: functions/src/controllers/campaign.js (hunk 1 above)
Comment (judeokafor) [reply functions/src/controllers/campaign.js]: no mixture of .then and await

**Review by judeokafor — APPROVED** [review]
Fix the await mixture with .then

**Pull request merged**